texastribune/salesforce-stripe
# Add support for a referral ID

#### What's this PR do?

Adds support for a referral ID. If a query param of `referralId` is specified then it will be copied to an opportunity or RDO generated from that page. 

#### Why are we doing this? How does it help us?

It allows us to ask donors to become aggregators and monitor how much they've raised on our behalf. We can then do various things with that -- give them gifts, enter them in contests, etc.

#### How should this be manually tested?

Add a query param of `referralId` to the URL of any donation page. Try both a bogus one and a real one. If you're using the regular sandbox then this is a "real" one: `0035B00000KYqw5`. Try circle, blast, recurring and one-time donations. All should create valid Contacts/Accounts/Opportunities/RDOs, even when the ID is invalid. You can watch the server logs to see the interactions. Note that you'll need to turn on Celery for this to work. Ask me to verify in the sandbox if you'd like. 

#### Have automated tests been added?

They've been updated.

#### Has this been tested on mobile?

No.

#### Are there performance implications?

No.

#### What are the relevant tickets?

https://3.basecamp.com/3098728/buckets/736178/todos/1215247103

#### How should this change be communicated to end users?

Just let Amanda know it's ready. 

#### Next steps?

Make sure the new fields have been created in the production instance.

#### Smells?

I'm not super excited about the retry behavior. Now that we've got multiple values that may fail because they're non-existent the code is more complex.
